File: src/services/tankCalculations.ts
import { PropertyValue, Tank, Movement, getEffectiveVolume } from '@/types';

/**
 * Calculate volume-weighted average properties when blending oils
 */
export function calculateBlendedProperties(
  tankVolume: number,
  tankProperties: PropertyValue[],
  addedVolume: number,
  addedProperties: PropertyValue[]
): PropertyValue[] {
  const totalVolume = tankVolume + addedVolume;

  if (totalVolume === 0) {
    return addedProperties.length > 0 ? addedProperties : tankProperties;
  }

  // Create a map of all property IDs from both sources
  const allPropertyIds = new Set([
    ...tankProperties.map((p) => p.propertyId),
    ...addedProperties.map((p) => p.propertyId),
  ]);

  const blendedProperties: PropertyValue[] = [];

  for (const propertyId of allPropertyIds) {
    const tankProp = tankProperties.find((p) => p.propertyId === propertyId);
    const addedProp = addedProperties.find((p) => p.propertyId === propertyId);

    const tankValue = tankProp?.value ?? null;
    const addedValue = addedProp?.value ?? null;

    let blendedValue: number | null;

    if (tankValue !== null && addedValue !== null) {
      // Both have values - calculate weighted average
      blendedValue =
        (tankVolume * tankValue + addedVolume * addedValue) / totalVolume;
    } else if (tankValue !== null && tankVolume > 0) {
      // Only tank has value - adjust proportion
      blendedValue = (tankVolume * tankValue) / totalVolume;
    } else if (addedValue !== null && addedVolume > 0) {
      // Only added has value - adjust proportion
      blendedValue = (addedVolume * addedValue) / totalVolume;
    } else {
      blendedValue = null;
    }

    blendedProperties.push({
      propertyId,
      value: blendedValue !== null ? Math.round(blendedValue * 1000) / 1000 : null,
    });
  }

  return blendedProperties;
}

/**
 * Calculate projected tank state based on scheduled movements
 */
export function calculateProjectedState(
  tank: Tank,
  scheduledMovements: Movement[]
): { volume: number; properties: PropertyValue[] } {
  let projectedVolume = tank.currentVolume;
  let projectedProperties = [...tank.properties];

  // Sort movements by scheduled date
  const sortedMovements = [...scheduledMovements].sort(
    (a, b) => new Date(a.scheduledDate).getTime() - new Date(b.scheduledDate).getTime()
  );

  for (const movement of sortedMovements) {
    const volume = getEffectiveVolume(movement);

    switch (movement.type) {
      case 'receive':
        if (movement.destinationTankId === tank.id) {
          // Add volume and blend properties
          projectedProperties = calculateBlendedProperties(
            projectedVolume,
            projectedProperties,
            volume,
            movement.properties
          );
          projectedVolume += volume;
        }
        break;

      case 'ship':
        if (movement.sourceTankId === tank.id) {
          // Remove volume (properties stay same)
          projectedVolume = Math.max(0, projectedVolume - volume);
        }
        break;

      case 'transfer':
        if (movement.sourceTankId === tank.id) {
          // Remove volume from this tank
          projectedVolume = Math.max(0, projectedVolume - volume);
        } else if (movement.destinationTankId === tank.id) {
          // Add volume to this tank with blended properties
          projectedProperties = calculateBlendedProperties(
            projectedVolume,
            projectedProperties,
            volume,
            movement.properties
          );
          projectedVolume += volume;
        }
        break;
    }
  }

  return {
    volume: Math.round(projectedVolume * 1000) / 1000,
    properties: projectedProperties,
  };
}

/**
 * Get volume change from a movement for a specific tank
 */
export function getVolumeChange(movement: Movement, tankId: string): number {
  const volume = getEffectiveVolume(movement);

  switch (movement.type) {
    case 'receive':
      return movement.destinationTankId === tankId ? volume : 0;
    case 'ship':
      return movement.sourceTankId === tankId ? -volume : 0;
    case 'transfer':
      if (movement.sourceTankId === tankId) return -volume;
      if (movement.destinationTankId === tankId) return volume;
      return 0;
    default:
      return 0;
  }
}

File: backend/app/services/tank_calculations.py
from typing import List, Dict, Any

from app.models.common import PropertyValue


def get_effective_volume(movement: Dict[str, Any]) -> float:
    """Get actual volume if available, otherwise expected volume."""
    actual = movement.get("actualVolume")
    return actual if actual is not None else movement["expectedVolume"]


def calculate_blended_properties(
    tank_volume: float,
    tank_properties: List[PropertyValue],
    added_volume: float,
    added_properties: List[PropertyValue],
) -> List[PropertyValue]:
    """Calculate volume-weighted average properties when blending oils."""
    total_volume = tank_volume + added_volume

    if total_volume == 0:
        return added_properties if added_properties else tank_properties

    # Create maps for quick lookup
    tank_props_map: Dict[str, float | None] = {
        p.propertyId: p.value for p in tank_properties
    }
    added_props_map: Dict[str, float | None] = {
        p.propertyId: p.value for p in added_properties
    }

    # Get all unique property IDs
    all_property_ids = set(tank_props_map.keys()) | set(added_props_map.keys())

    blended_properties: List[PropertyValue] = []

    for prop_id in all_property_ids:
        tank_value = tank_props_map.get(prop_id)
        added_value = added_props_map.get(prop_id)

        blended_value: float | None = None

        if tank_value is not None and added_value is not None:
            # Both have values - weighted average
            blended_value = (
                tank_volume * tank_value + added_volume * added_value
            ) / total_volume
        elif tank_value is not None and tank_volume > 0:
            # Only tank has value
            blended_value = (tank_volume * tank_value) / total_volume
        elif added_value is not None and added_volume > 0:
            # Only added has value
            blended_value = (added_volume * added_value) / total_volume

        blended_properties.append(
            PropertyValue(
                propertyId=prop_id,
                value=round(blended_value, 3) if blended_value is not None else None,
            )
        )

    return blended_properties


def calculate_projected_state(
    tank: Dict[str, Any], scheduled_movements: List[Dict[str, Any]]
) -> Dict[str, Any]:
    """Calculate projected tank state based on scheduled movements."""
    projected_volume = tank["currentVolume"]
    projected_properties = [PropertyValue(**p) for p in tank["properties"]]

    # Sort movements by scheduled date
    sorted_movements = sorted(scheduled_movements, key=lambda m: m["scheduledDate"])

    for movement in sorted_movements:
        volume = get_effective_volume(movement)
        movement_type = movement["type"]
        movement_props = [PropertyValue(**p) for p in movement.get("properties", [])]

        if movement_type == "receive" and movement.get("destinationTankId") == tank["id"]:
            projected_properties = calculate_blended_properties(
                projected_volume, projected_properties, volume, movement_props
            )
            projected_volume += volume

        elif movement_type == "ship" and movement.get("sourceTankId") == tank["id"]:
            projected_volume = max(0, projected_volume - volume)

        elif movement_type == "transfer":
            if movement.get("sourceTankId") == tank["id"]:
                projected_volume = max(0, projected_volume - volume)
            elif movement.get("destinationTankId") == tank["id"]:
                projected_properties = calculate_blended_properties(
                    projected_volume, projected_properties, volume, movement_props
                )
                projected_volume += volume

    return {
        "volume": round(projected_volume, 3),
        "properties": projected_properties,
    }


def get_volume_change(movement: Dict[str, Any], tank_id: str) -> float:
    """Get signed volume change for a tank from a movement."""
    volume = get_effective_volume(movement)
    movement_type = movement["type"]

    if movement_type == "receive":
        return volume if movement.get("destinationTankId") == tank_id else 0
    elif movement_type == "ship":
        return -volume if movement.get("sourceTankId") == tank_id else 0
    elif movement_type == "transfer":
        if movement.get("sourceTankId") == tank_id:
            return -volume
        if movement.get("destinationTankId") == tank_id:
            return volume
    return 0
